sos/collector/clusters/juju.py

```diff
@@ -0,0 +1,174 @@
+            for parent in subordinate_to:
+                index["apps"][app].extend(index["apps"][parent])
+                units = juju_status["applications"][parent]["units"]
+                for unit, unit_info in units.items():
+                    node = f"{model_name}:{unit_info['machine']}"
+                    for sub_key, sub_value in unit_info.get(
+                        "subordinates", {}
+                    ).items():
+                        if sub_key.startswith(app + "/"):
+                            index["units"][sub_key] = [node]
+
+    def _add_machines_to_index(self, index, juju_status, model_name):
+        """Add machines to index.
+
+        If model does not have any applications it needs to be manually added.
+        """
+        for machine in juju_status["machines"].keys():
+            node = f"{model_name}:{machine}"
+            index["machines"][machine] = [node]
+
+    def _execute_juju_status(self, model_name):
+        model_option = f"-m {model_name}" if model_name else ""
+        format_option = "--format json"
+        status_cmd = f"{self.cmd} status {model_option} {format_option}"
+        juju_status = None
+        res = self.exec_primary_cmd(status_cmd)
+        if res["status"] == 0:
+            juju_status = json.loads(
+                self._cleanup_juju_output((res["output"]))
+            )
+        else:
+            raise Exception(f"{status_cmd} did not return usable output")
+        return juju_status
+
+    def _filter_by_pattern(self, key, patterns, model_info):
+        """Filter with regex match."""
+        nodes = set()
+        for pattern in patterns:
+            for param, value in model_info[key].items():
+                if re.match(pattern, param):
+                    nodes.update(value or [])
+        return nodes
+
+    def _filter_by_fixed(self, key, patterns, model_info):
+        """Filter with fixed match."""
+        nodes = set()
+        for pattern in patterns:
+            for param, value in model_info[key].items():
+                if pattern == param:
+                    nodes.update(value or [])
+        return nodes
+
+    def set_transport_type(self):
+        """Dynamically change transport to 'juju'."""
+        return "juju"
+
+    def get_nodes(self):
+        """Get the machine numbers from `juju status`."""
+        models = _parse_option_string(self.get_option("models"))
+        apps = _parse_option_string(self.get_option("apps"))
+        units = _parse_option_string(self.get_option("units"))
+        machines = _parse_option_string(self.get_option("machines"))
+        filters = {"apps": apps, "units": units, "machines": machines}
+
+        if not models:
+            models = [""]  # use current model by default
+
+        nodes = set()
+
+        for model in models:
+            model_info = self._get_model_info(model)
+            if not any(filters.values()):
+                pass
```

Comment:
"if not any(filters.values())" then this method returns `[]` - does it make sense to move this test before the `for model in models` loop to skip calling `_get_model_info` / whole cycle?